src/vs/workbench/contrib/chat/browser/promptFileReference.ts

```diff
@@ -0,0 +1,393 @@
+			return;
+		}
+
+		// if file is changed or deleted, re-resolve the file reference
+		// in the case when the file is deleted, this should result in
+		// failure to open the file, so the `errorCondition` field will
+		// be updated to an appropriate error instance and the `children`
+		// field will be cleared up
+		this.resolve();
+	}
+
+	/**
+	 * Get file stream, if the file exsists.
+	 */
+	private async getFileStream(): Promise<IFileStreamContent | null> {
+		// if URI doesn't point to a prompt snippet file, don't try to resolve it
+		if (this.uri.path.endsWith(PROMP_SNIPPET_FILE_EXTENSION) === false) {
+			this._errorCondition = new NotPromptSnippetFile(this.uri);
+
+			return null;
+		}
+
+		try {
+			return await this.fileService.readFileStream(this.uri);
+		} catch (error) {
+			this._errorCondition = new FileOpenFailed(this.uri, error);
+
+			return null;
+		}
+	}
+
+	/**
+	 * Resolve the current file reference on the disk and
+	 * all nested file references that may exist in the file.
+	 *
+	 * @param waitForChildren Whether need to block until all child references are resolved.
+	 */
+	public async resolve(
+		waitForChildren: boolean = false,
+	): Promise<this> {
+		return await this.resolveReference(waitForChildren);
+	}
+
+	/**
+	 * Private implementation of the {@link resolve} method, that allows
+	 * to pass `seenReferences` list to the recursive calls to prevent
+	 * infinite file reference recursion.
+	 */
+	private async resolveReference(
+		waitForChildren: boolean = false,
+		seenReferences: string[] = [],
+	): Promise<this> {
+		// remove current error condition from the previous resolve attempt, if any
+		delete this._errorCondition;
+
+		// dispose current child references, if any exist from a previous resolve
+		this.disposeChildren();
+
+		// to prevent infinite file recursion, we keep track of all references in
+		// the current branch of the file reference tree and check if the current
+		// file reference has been already seen before
+		if (seenReferences.includes(this.uri.path)) {
+			seenReferences.push(this.uri.path);
+
+			this._errorCondition = new RecursiveReference(this.uri, seenReferences);
+			this._resolveAttempted = true;
+			this._onUpdate.fire();
+
+			return this;
+		}
+
+		// we don't care if reading the file fails below, hence can add the path
+		// of the current reference to the `seenReferences` set immediately, -
+		// even if the file doesn't exist, we would never end up in the recursion
+		seenReferences.push(this.uri.path);
+
+		// try to get stream for the contents of the file, it may
+		// fail to multiple reasons, e.g. file doesn't exist, etc.
+		const fileStream = await this.getFileStream();
+		this._resolveAttempted = true;
+
+		// failed to open the file, nothing to resolve
+		if (fileStream === null) {
+			this._onUpdate.fire();
+
+			return this;
+		}
+
+		// get all file references in the file contents
+		const references = await this.codec.decode(fileStream.value).consumeAll();
```

Comment:
Just wondering whether this needs to be a stream? Do you use it as a stream?

Reply:
Maybe it's just nice to have the stream since you are working from a file stream, just wondering if there is a case where you would use it that way

Reply:
Because the amount of data we need to parse might be substantial and because filesystem IO is involved, I thought it is important to keep it as a stream so we can do the parsing concurrently.

I have a follow-up task to make the logic here take leverage of that. Beside this line, the decoders are composed(e.g., `PromptFileDecoder` uses `SimpleTokenDecoder`) hence fully take advantage of the fact that an underlying decoder is a stream of messages 🚀